From 20fc0873a96276af045928ffac621f79be74eb5b Mon Sep 17 00:00:00 2001
From: Paul Price <price@astro.princeton.edu>
Date: Tue, 15 Aug 2017 12:03:40 -0700
Subject: [PATCH] fix false compression error

Could get 'Lossless compression of floating point images' error even when
compressing integer images. Check the type before throwing the error.
---
 imcompress.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/imcompress.c b/imcompress.c
index 78da8ed..a2baca8 100644
--- a/imcompress.c
+++ b/imcompress.c
@@ -969,7 +969,7 @@ int imcomp_init_table(fitsfile *outfptr,
 
     /* check for special case of losslessly compressing floating point */
     /* images.  Only compression algorithm that supports this is GZIP */
-    if ( (outfptr->Fptr)->request_quantize_level == NO_QUANTIZE) {
+    if (inbitpix < 0 && (outfptr->Fptr)->request_quantize_level == NO_QUANTIZE) {
        if (((outfptr->Fptr)->request_compress_type != GZIP_1) &&
            ((outfptr->Fptr)->request_compress_type != GZIP_2)) {
          ffpmsg("Lossless compression of floating point images must use GZIP (imcomp_init_table)");
@@ -1720,8 +1720,16 @@ int imcomp_compress_tile (fitsfile *outfptr,
     if ( (outfptr->Fptr)->quantize_level == NO_QUANTIZE) {
        if (((outfptr->Fptr)->compress_type != GZIP_1) &&
            ((outfptr->Fptr)->compress_type != GZIP_2)) {
-         ffpmsg("Lossless compression of floating point images must use GZIP (imcomp_compress_tile)");
-         return(*status = DATA_COMPRESSION_ERR);
+          switch (datatype) {
+             case TFLOAT:
+             case TDOUBLE:
+             case TCOMPLEX:
+             case TDBLCOMPLEX:
+                ffpmsg("Lossless compression of floating point images must use GZIP (imcomp_compress_tile)");
+                return(*status = DATA_COMPRESSION_ERR);
+             default:
+                break;
+          }
        }
     }
 
-- 
2.13.4

